shell/common/shell.cc

```diff
@@ -1246,6 +1257,12 @@ bool Shell::HandleServiceProtocolMessage(
 // |ServiceProtocol::Handler|
 ServiceProtocol::Handler::Description Shell::GetServiceProtocolDescription()
     const {
+  FML_DCHECK(task_runners_.GetUITaskRunner()->RunsTasksOnCurrentThread());
+
+  if (!engine_) {
```

Comment:
It's a bit obscure. For the case 'ShellTest.OnServiceProtocolGetSkSLsWorks', `Shell` is created and destroyed without running the engine. When destructing `Shell`, `engine_` was moved in `Shell::~Shell()` https://github.com/flutter/engine/blob/master/shell/common/shell.cc#L395, and then comes `Shell::OnRootIsolateCreated()` and calls `GetServiceProtocolDescription()` meeting the invalid `engine_`.
Anyway, it is a protection better than none.

Reply:
We reviewed this patch again and I believe this is good to go except this one case where we all preferred if it were an assert. The unit test seems to be accessing the private method because it is a friend of the shell. Now that this code has been reworked, the assumption in the test failed. Can we rework the test to not make this assumption anymore? Let's land then patch then.

Reply:
Appreciate and agree your comments. To keep this as an assert, I just simply change `engine_` to `weak_engine_` since it should run on the ui thread. For the test case, when it was called from `Shell::OnRootIsolateCreated()`, `engine_` had been moved but not destructed and `weak_engine_` was safe to refer.

Reply:
I've proposed a PR that starts an engine in `ShellTest.OnServiceProtocolGetSkSLsWorks` (https://github.com/flutter/engine/pull/20801).

With that change, a workaround in `Shell::GetServiceProtocolDescription ` will not be needed.

Reply:
I've patched #20801 and run the test locally without the workaround in `Shell::GetServiceProtocolDescription`, and the result is crash for the same reason that `Shell::~Shell()` is called before `Shell::OnRootIsolateCreated()`. I think using `weak_engine_` instead of  `engine_` inside `Shell::GetServiceProtocolDescription` is more reasonable.

Reply:
Ah, now I see - the `Shell::~Shell()` call will happen on the platform thread and will transfer ownership of the shell's `engine_` into the task queued to the UI thread that deletes the engine.

The task that calls `Shell::OnRootIsolateCreated` was previously queued to the UI thread when `CreateShellOnPlatformThread` posted the task that creates the `Engine` and its `RuntimeController`.  So when `Shell::OnRootIsolateCreated` calls `GetServiceProtocolDescription`, the `Engine` deletion task has not yet run and it's safe to assume that `weak_engine_` is still valid.

Reply:
Yeah, that it is.

Reply:
I'm glad we figured it out. Maybe put Jason's reasoning in the comment so future contributors would understand why we chose `weak_engine_` here instead of `engine_`.

Alternatively, there are many pieces of `if (!weak_engine_) {...}` or `[engine = weak_engine_]{ if (engine) { ... } }` code in `shell.cc`. So it looks very reasonable if we put a `if (!weak_engine_) {...}` here.

@jason-simmons : there's one more thing that I don't completely understand. Line [1641][] won't return until the UI thread task to run the engine was fully completed. The shell destruction happened on line [1651][] after it. The UI thread task of `OnRootIsolateCreated` was added at `RuntimeController` and `Engine` construction which seems to be earlier than `RunEngine`. So I'd expect `OnRootIsolateCreated` to be executed before shell destruction in line [1651][]?

[1641]: https://github.com/flutter/engine/pull/20801/files#diff-21bc42b7f14c143ec0333cc2e1212807R1641
[1651]: https://github.com/flutter/engine/pull/20801/files#diff-21bc42b7f14c143ec0333cc2e1212807R1651

Reply:
The crash in the unit test happened because `Shell::engine_` was null even though the shell had not yet been fully destructed and the engine had not been deleted.

`Shell::engine_` is set to null on the platform thread within the Shell destructor by https://github.com/flutter/engine/blob/master/shell/common/shell.cc#L395.  The platform thread can do this at any time regardless of what the UI thread is doing.  The platform thread will then queue a task to the UI thread that actually deletes the engine.

So it's unsafe to access `Shell::engine_` directly from a UI thread task.  UI thread tasks should use `weak_engine_` and check if it is null.